File: interview-preparation-kit/archive/searching/swap_nodes_algo.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({input: process.stdin, output: process.stdout});

let inputString = [];

rl.prompt();

rl.on('line', (inputLine) => {
    inputString.push(inputLine);
})

rl.on('close', () => {
    console.log(main(inputString));
    process.exit(0);
})

function dfsInorderSwapNodes(tree, swapDepthMultiples) {
    let arr = Array();
    for (let i=0; i<swapDepthMultiples.length; i++) {
        const swapDepthMultiple = swapDepthMultiples[i]
        let inorder = Array();
        let stack = Array();
        let currentIndex = 1;
        let depth = 0
        while (true) {
            if (currentIndex!=-1) {
                depth+=1
                if (depth%swapDepthMultiple==0) {
                    const leftNode = tree[currentIndex-1][0]
                    tree[currentIndex-1][0] = tree[currentIndex-1][1]
                    tree[currentIndex-1][1] = leftNode
                }
                stack.push([currentIndex, depth])
                currentIndex = tree[currentIndex-1][0]
            } else if (stack.length>0) {
                const currentIndexAndDepth = stack.pop()
                depth = currentIndexAndDepth[1]
                inorder.push([currentIndexAndDepth[0], depth])
                currentIndex = tree[currentIndexAndDepth[0]-1][1]
            } else {
                break
            }
        }
        arr.push(inorder.map(elem => elem[0]))
    }
    return arr
}

function main(inputString) {
    // return inputString
    const numNodes = inputString[0];

    let tree = Array(numNodes);
    for (let row=1; row<=numNodes; row++) {
        tree[row-1] = inputString[row].split(' ').map(node => parseInt(node, 10));
    }

    const numSwapsIndex = 1+tree.length
    const numSwaps = parseInt(inputString[numSwapsIndex], 10)

    let swapDepthMultiples = [];
    for (let i=numSwapsIndex+1; i<inputString.length; i++) {
        swapDepthMultiples.push(parseInt(inputString[i], 10));
    }
    console.log(swapDepthMultiples)

    let result = dfsInorderSwapNodes(tree, swapDepthMultiples);
    return result
}

function swapNodes(indexes, queries) {
  const swap = (k, node, depth) => {
    if (!node) {
      return;
    }
    swap(k, node.left, depth + 1);
    swap(k, node.right, depth + 1);
    if (depth % k === 0) {
      const temp = node.left;
      node.left = node.right;
      node.right = temp;
    }
  }

  const inOrderTraversal = (order, node) => {
    if (!node) {
      return;
    }
    inOrderTraversal(order, node.left)
    order.push(node.value);
    inOrderTraversal(order, node.right);
  }

  const result = [];
  const indexesLength = indexes.length;
  const treeNodes = Array(indexesLength + 1);
  for (let i = 1; i <= indexesLength; i += 1) {
    if (!treeNodes[i]) {
      treeNodes[i] = { value: i, left: null, right: null };
    }
    const leftValue = indexes[i - 1][0];
    if (leftValue !== -1) {
      if (!treeNodes[leftValue]) {
        treeNodes[leftValue] = { value: leftValue, left: null, right: null };
      }
      treeNodes[i].left = treeNodes[leftValue];
    }
    const rightValue = indexes[i - 1][1];
    if (rightValue !== -1) {
      if (!treeNodes[rightValue]) {
        treeNodes[rightValue] = { value: rightValue, left: null, right: null };
      }
      treeNodes[i].right = treeNodes[rightValue];
    }
  }

  for (let j = 0; j < queries.length; j += 1) {
    const k = queries[j];
    swap(k, treeNodes[1], 1);
    const order = [];
    inOrderTraversal(order, treeNodes[1]);
    result.push(order);
  }
  return result;
}

File: interview-preparation-kit/archive/searching/swap_nodes_algo.py
#!/bin/python3
import os
import sys

def dfs_inorder_swap_nodes(tree, swap_depth_multiples):
    arr = []
    for swap_depth_multiple in swap_depth_multiples:
        inorder = []
        stack = []
        current_index = 1
        depth = 0
        while True:
            #Note you can't set current_node = tree[current_index-1] here because
            #you'll reach current_index=-1 eventally so things will break!
            if current_index!=-1: #Keep traversing until you reach -1
                depth+=1 #Note this is the only place we add incrementall to depth
                if depth%swap_depth_multiple==0: #restructure the tree?
                    # left_node = tree[current_index-1][0]
                    # tree[current_index-1][0] = tree[current_index-1][1]
                    # tree[current_index-1][1] = left_node
                    tree[current_index-1][0], tree[current_index-1][1] = tree[current_index-1][1], tree[current_index-1][0]
                stack.append((current_index, depth))
                current_index = tree[current_index-1][0]
            elif stack:
                #When the current index is eventually -1, we will start having to move up,
                #and the order of elements in the stack IS THE inorder order
                current_index_depth = stack.pop()
                depth = current_index_depth[1]
                #The depth needs to be reset to whichever node we are current at as we start moving back up,
                #because if there is a node on the right tree, we must add 1
                inorder.append(current_index_depth)
                current_index = tree[current_index_depth[0]-1][1]
            else:
                break
        arr.append(list(list(zip(*inorder))[0]))
    return arr

if __name__ == '__main__':
    # indexes = [[2, 3], [4, 5], [6, -1], [7, -1], [8, -1], [9, 10], [-1, 11],
        # [-1, -1], [-1, -1], [-1, -1], [-1, -1]] #My customer tree with greater complexity to chekc if nodes on the right were being correctly accounted!
    indexes = [[2, 3], [4, -1], [5, -1], [6, -1], [7, 8], [-1, 9], [-1, -1],
        [10, 11], [-1, -1], [-1, -1], [-1, -1]]

    queries = [2, 4]

    result = dfs_inorder_swap_nodes(indexes, queries)
    print(result)
    # fptr.write('\n'.join([' '.join(map(str, x)) for x in result]))
    # fptr.write('\n')

    # fptr.close()
